### ISSUE TITLE
Dependency on external package causes compiler preferences to be ignored

### ISSUE THREAD

**Issue by sleak-lbl** [LLNL/spack#21311]
### Steps to reproduce the issue

In my `packages.yaml` I have:

```
packages:
  all:
    compiler: [intel@19.0.3.199, gcc@8.3.0]
[snip]
  cmake:
    externals:
    - spec: cmake@3.14.4%intel
      modules:
      - cmake/3.14.4
    - spec: cmake@3.14.4%cce
      modules:
      - cmake/3.14.4
    - spec: cmake@3.14.4%gcc
      modules:
      - cmake/3.14.4
```

When I run `spack spec -Il clingo@master` I expect to see the intel compiler being used, since that is specified as the preferred compiler, eg:

```
 -   nsxzgyg  clingo@master%intel@19.0.3.199~docs~ipo+python build_type=RelWithDebInfo arch=cray-cnl7-haswell
[snip]
```

However, I get:

```
 -   t7h77vt  clingo@master%cce@9.1.3~docs~ipo+python build_type=RelWithDebInfo arch=cray-cnl7-haswell
 [snip]
```

If I remove the entries for cmake as an external package, I get the expected result.

### Apparent Cause

After instrumenting some of the concretizing routines, I think the key is at `lib/spack/spack/concretize.py:choose_virtual_or_external`  .. it sorts the candidates by abi compatibility, but doesn’t consider the preferred compiler as set in `packages.yaml`. The reason cce gets chosen is simply that it comes earlier alphabetically than gcc or intel.

### Information on your system

```
20:55 sleak@cori03:reproducer$ spack debug report
* **Spack:** 0.16.0
* **Python:** 3.6.10
* **Platform:** cray-sles15-haswell
```

### Additional information

- [x] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [x] I have searched the issues of this repo and believe this is not a duplicate
- [ ] I have run the failing commands in debug mode and reported the output

**Comment by alalazo** [comment]
Am I correct that this issue is only with the original concretizer and not with the ASP based concretizer? Is it by any chance it solved by #20976 (since it looks similar to that issue)?

**Comment by alalazo** [comment]
Closing, since there has been no activity in a while. Feel free to reopen if you want to continue work on this.
